src/gui/transferlistsortmodel.cpp

```diff
@@ -86,179 +86,157 @@ void TransferListSortModel::disableTrackerFilter()
-            const bool isActiveL = TorrentFilter::ActiveTorrent.match(model->torrentHandle(left));
-            const bool isActiveR = TorrentFilter::ActiveTorrent.match(model->torrentHandle(right));
-            if (isActiveL != isActiveR)
-                return isActiveL;
-
-            const auto queuePosL = left.sibling(left.row(), TransferListModel::TR_QUEUE_POSITION)
-                    .data(TransferListModel::UnderlyingDataRole).toInt();
-            const auto queuePosR = right.sibling(right.row(), TransferListModel::TR_QUEUE_POSITION)
-                    .data(TransferListModel::UnderlyingDataRole).toInt();
-            const bool isSeedingL = (queuePosL < 0);
-            const bool isSeedingR = (queuePosR < 0);
-            if (isSeedingL != isSeedingR)
-            {
-                const bool isAscendingOrder = (sortOrder() == Qt::AscendingOrder);
-                if (isSeedingL)
-                    return !isAscendingOrder;
-
-                return isAscendingOrder;
-            }
-
-            const auto etaL = leftValue.toLongLong();
-            const auto etaR = rightValue.toLongLong();
-            const bool isInvalidL = ((etaL < 0) || (etaL >= MAX_ETA));
-            const bool isInvalidR = ((etaR < 0) || (etaR >= MAX_ETA));
-            if (isInvalidL && isInvalidR)
-            {
-                if (isSeedingL)  // Both seeding
-                    return invokeLessThanForColumn(TransferListModel::TR_SEED_DATE);
-
-                return (queuePosL < queuePosR);
-            }
-
-            if (!isInvalidL && !isInvalidR)
-                return (etaL < etaR);
-
-            return !isInvalidL;
-        }
 
-    case TransferListModel::TR_LAST_ACTIVITY:
-        {
-            const auto lastActivityL = leftValue.toLongLong();
-            const auto lastActivityR = rightValue.toLongLong();
-
-            if (lastActivityL < 0) return false;
-            if (lastActivityR < 0) return true;
-            return lastActivityL < lastActivityR;
+            if ((totalL < 0) && (totalR < 0))
+                break;
+            else if (totalR < 0)
+                return true;
+            else if (totalL < 0)
+                return false;
+            else if (totalL != totalR)
+                return totalL < totalR;
         }
+        break;
 
-    case TransferListModel::TR_RATIO_LIMIT:
-        {
-            const auto ratioL = leftValue.toReal();
-            const auto ratioR = rightValue.toReal();
-
-            if (ratioL < 0) return false;
-            if (ratioR < 0) return true;
-            return ratioL < ratioR;
-        }
+    default:
+        Q_ASSERT(false);
+        break;
     }
 
-    return (leftValue != rightValue)
-        ? QSortFilterProxyModel::lessThan(left, right)
-        : invokeLessThanForColumn(TransferListModel::TR_QUEUE_POSITION);
+    // Since the values are equal, compare by original order. This is to
```

Comment:
Isn't `return false;` enough?

Reply:
No. This is to main core of retaining the stable sort property, without it everything will fall apart.

Reply:
>No. This is to main core of retaining the stable sort property, without it everything will fall apart.

Really? Did you test it?
Strange, I was sure that the usual sorting algorithms preserve the order of the similar elements. Otherwise, it looks too wasteful.

Reply:
>Really? 

Which part of the comment are you referring to?

>Did you test it?

Yes, you might want to download the test build to see it yourself.

>Strange, I was sure that the usual sorting algorithms preserve the order of the similar elements. Otherwise, it looks too wasteful.

I always considered Qt implemented in a weird way such that stable sorting is impossible until I figured it out.
IMO only relying on `lessThan()` isn't enough to maintain a stable sorting, it needs to detect the case when the comparison is equal and then act on it.

Reply:
Anyway I found it is possible to simplify further, PR updated.

Reply:
>Which part of the comment are you referring to?

Quoted one:
>This is to main core of retaining the stable sort property, without it everything will fall apart.

I am surprised that Qt sorting is unstable with respect to the "similar" elements. This is illogical and irrational (unless of course it uses some advanced multithreading method, which I strongly doubt).

Reply:
>I am surprised that Qt sorting is unstable with respect to the "similar" elements. 

I did some testing years ago and that was my conclusion at the time, seems it is different now.